src/vision/status_quo/alan_finds_database_drops_hard.md

```diff
@@ -0,0 +1,104 @@
+
+
+## Searching for the Solution
+
+
+Alan tries to figure out what happened from the logs, but the only information he sees is that a new connection has been received. Alan turns to the documentation for the `sqlx` crate to see if there are flags that might enable extra instrumentation but he can't find any [sqlx::Connection::close](https://docs.rs/sqlx/0.5.1/sqlx/trait.Connection.html#required-methods).
+
+* He does find the [`close` method] which mentions "This method is not required for safe and consistent operation. However, it is recommended to call it instead of letting a connection drop as the database backend will be faster at cleaning up resources."
+* He adds a call to `close` into his code and it helps some but he is still able to reproduce the problem if he refreshes often enough. 
+* He adds a log statement right before calling `close` to see if it is working:
+
+```rust
+use sqlx::Connection;
+
+#[async_std::main]
+async fn do_the_thing() -> Result<(), sqlx::Error> {
+    // Create a connection
+    let conn = SqliteConnection::connect("sqlite::memory:").await?;
+
+    // Make a simple query to return the given parameter
+    let row: (i64,) = sqlx::query_as("SELECT $1")
+        .bind(150_i64)
+        .fetch_one(&conn).await?; // <----- if this await is cancelled, doesn't help
+
+    assert_eq!(row.0, 150);
+    
+    // he adds this:
+    log!("closing the connection");
+    conn.close();
+
+    Ok(())
+}
+```
+  
+* He observes that in the cases where he has the problem the log statement never executes.
+* He asks Barbara for help and she explains how `await` can be canceled and it will the destructors for things that are in scope.
+* He reads the [source for the SqliteConnection destructor](https://github.com/launchbadge/sqlx/blob/0ed524d65c2a3ee2e2a6706910b85bf2bb72115f/sqlx-core/src/pool/connection.rs#L70-L74) and finds that destructor spawns a task to actually close the connection.
+* He realizes there is a race condition:
+    * the task may not have actually closed the connection before `do_the_thing` is called a second time
+* He gives up and gets frustrated
+
+Next, Alan seeks verification and validation of his understanding of the source code from the `sqlx` forum. His advisor states that Rust doesn't have a way to execute async operations in a destructor.
+
+## Finding the Solution
+
+Alan briefly considers rearchitecting his application in more extreme ways to retain use of async, but he gives up and seeks a more straight forward solution. He discovers `rusqlite`, a sychronous database library and adopts it. This requires some rearchitecting but solves the problem.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+* **What are the morals of the story?**
+    * The story demonstrates solid research steps that Alan uses to understand and resolve his problem.
```

Comment:
I'd say a key moral is that Rust's async story is lacking a way to executing async operations in destructors. Spawning is a workaround, but it can have unexpected side-effects.

Reply:
added